### PR TITLE
Introduced vectorized batch norm backward kernels

### PR DESCRIPTION
**Pull request by xytintel**
Add vectorization implementations of batch norm backward kernels, with speeding up by up to 1.27x for channels-last cases and 1.32x for normal cases.

### CODE REVIEW

**Comment by xytintel** [comment]
Test script:
```python
N, C, H, W = 256, 256, 56, 56
x = torch.randn(N, C, H, W).xpu().to(memory_format=torch.channels_last).bfloat16()
x.requires_grad = True
bn = nn.BatchNorm2d(C).xpu()

prof_xpu = torch.profiler.profile(
    activities=[
    torch.profiler.ProfilerActivity.CPU,
    torch.profiler.ProfilerActivity.XPU],
)
with prof_xpu:
    for i in range(128):
        output = bn(x)
        output.backward(x)
print(prof_xpu.key_averages(group_by_input_shape=True).table(sort_by="self_xpu_time_total", row_limit=100000))
print(output.dtype)
```

Old:
```txt
# -------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
#                                                    Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg      Self XPU    Self XPU %     XPU total  XPU time avg    # of Calls  
# -------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
#                                             aten::fill_        19.37%     106.887ms        30.11%     166.160ms     216.354us      88.489ms        40.01%      88.489ms     115.220us           768  
# at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us      86.077ms        38.92%      86.077ms     551.777us           156  
#                                              aten::add_        19.89%     109.761ms        27.48%     151.666ms     297.968us      65.633ms        29.68%      65.633ms     128.946us           509  
#                                 aten::native_batch_norm        17.23%      95.108ms        29.56%     163.157ms       1.275ms      50.297ms        22.74%      51.549ms     402.727us           128  
# at::native::xpu::BatchNormTransformInputChannelsLast...         0.00%       0.000us         0.00%       0.000us       0.000us      43.496ms        19.67%      43.496ms       1.115ms            39  
# at::native::xpu::VectorizedElementwiseKernel<8, at::...         0.00%       0.000us         0.00%       0.000us       0.000us      40.185ms        18.17%      40.185ms       1.058ms            38  
# at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us      25.346ms        11.46%      25.346ms     342.519us            74  
#                        aten::native_batch_norm_backward        10.25%      56.574ms        39.25%     216.585ms       1.692ms      16.734ms         7.57%     103.971ms     812.273us           128  
# at::native::xpu::BatchNormBackwardElemtChannelsLastK...         0.00%       0.000us         0.00%       0.000us       0.000us       9.651ms         4.36%       9.651ms     247.458us            39  
# at::native::xpu::BatchNormBackwardReduceChannelsLast...         0.00%       0.000us         0.00%       0.000us       0.000us       7.083ms         3.20%       7.083ms     181.625us            39  
# at::native::xpu::WelfordBatchNormStatChannelsLastVec...         0.00%       0.000us         0.00%       0.000us       0.000us       5.505ms         2.49%       5.505ms     141.165us            39  
# at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       2.411ms         1.09%       2.411ms      30.915us            78  
# at::native::xpu::UnrolledElementwiseForMultiOutputsK...         0.00%       0.000us         0.00%       0.000us       0.000us       1.296ms         0.59%       1.296ms      33.218us            39  
# at::native::xpu::VectorizedElementwiseKernel<2, at::...         0.00%       0.000us         0.00%       0.000us       0.000us     101.600us         0.05%     101.600us       2.605us            39  
#                                   urEnqueueKernelLaunch        25.54%     140.943ms        25.54%     140.943ms      73.523us       0.000us         0.00%       0.000us       0.000us          1917  
#                                        aten::batch_norm         0.19%       1.026ms        30.37%     167.617ms       1.310ms       0.000us         0.00%      51.549ms     402.727us           128  
#                            aten::_batch_norm_impl_index         0.40%       2.211ms        30.19%     166.591ms       1.301ms       0.000us         0.00%      51.549ms     402.727us           128  
#                                             aten::empty         1.70%       9.357ms         1.70%       9.357ms       6.646us       0.000us         0.00%       0.000us       0.000us          1408  
#                                        aten::empty_like         0.32%       1.749ms         0.86%       4.747ms      18.544us       0.000us         0.00%       0.000us       0.000us           256  
#                                     aten::empty_strided         0.54%       2.998ms         0.54%       2.998ms      11.711us       0.000us         0.00%       0.000us       0.000us           256  
#                                             aten::zeros         1.05%       5.810ms        32.83%     181.189ms     235.923us       0.000us         0.00%      88.489ms     115.220us           768  
#                                             aten::zero_         0.62%       3.443ms        30.73%     169.603ms     220.837us       0.000us         0.00%      88.489ms     115.220us           768  
# autograd::engine::evaluate_function: NativeBatchNorm...         0.85%       4.709ms        40.51%     223.565ms       1.747ms       0.000us         0.00%     103.971ms     812.273us           128  
#                                NativeBatchNormBackward0         0.41%       2.272ms        39.66%     218.857ms       1.710ms       0.000us         0.00%     103.971ms     812.273us           128  
# autograd::engine::evaluate_function: torch::autograd...         0.78%       4.302ms        15.12%      83.422ms     217.244us       0.000us         0.00%      65.532ms     170.655us           384  
#                         torch::autograd::AccumulateGrad         0.85%       4.673ms        14.34%      79.120ms     206.041us       0.000us         0.00%      65.532ms     170.655us           384  
#                                            aten::detach         0.01%      29.475us         0.01%      53.308us      17.769us       0.000us         0.00%       0.000us       0.000us             3  
#                                                  detach         0.00%      23.833us         0.00%      23.833us       7.944us       0.000us         0.00%       0.000us       0.000us             3  
# -------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
# Self CPU time total: 551.876ms
# Self XPU time total: 221.153ms
```

Optimized:
```txt
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg      Self XPU    Self XPU %     XPU total  XPU time avg    # of Calls  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
                                            aten::fill_        18.75%      94.770ms        29.25%     147.848ms     192.510us      69.647ms        34.96%      69.647ms      90.686us           768  
at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us      67.200ms        33.73%      67.200ms     430.770us           156  
                                             aten::add_        19.77%      99.940ms        27.69%     139.940ms     274.931us      65.792ms        33.03%      65.792ms     129.258us           509  
                                aten::native_batch_norm        17.53%      88.606ms        30.08%     151.998ms       1.187ms      50.501ms        25.35%      51.789ms     404.605us           128  
at::native::xpu::BatchNormTransformInputChannelsLast...         0.00%       0.000us         0.00%       0.000us       0.000us      44.299ms        22.24%      44.299ms       1.136ms            39  
at::native::xpu::VectorizedElementwiseKernel<8, at::...         0.00%       0.000us         0.00%       0.000us       0.000us      40.301ms        20.23%      40.301ms       1.061ms            38  
at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us      25.391ms        12.75%      25.391ms     343.116us            74  
                       aten::native_batch_norm_backward        10.60%      53.552ms        38.75%     195.828ms       1.530ms      13.277ms         6.66%      81.636ms     637.779us           128  
at::native::xpu::BatchNormBackwardReduceChannelsLast...         0.00%       0.000us         0.00%       0.000us       0.000us       7.110ms         3.57%       7.110ms     182.306us            39  
at::native::xpu::BatchNormBackwardElemtChannelsLastV...         0.00%       0.000us         0.00%       0.000us       0.000us       6.167ms         3.10%       6.167ms     158.133us            39  
at::native::xpu::WelfordBatchNormStatChannelsLastVec...         0.00%       0.000us         0.00%       0.000us       0.000us       4.907ms         2.46%       4.907ms     125.822us            39  
at::native::xpu::VectorizedElementwiseKernel<4, at::...         0.00%       0.000us         0.00%       0.000us       0.000us       2.447ms         1.23%       2.447ms      31.366us            78  
at::native::xpu::UnrolledElementwiseForMultiOutputsK...         0.00%       0.000us         0.00%       0.000us       0.000us       1.295ms         0.65%       1.295ms      33.210us            39  
at::native::xpu::VectorizedElementwiseKernel<2, at::...         0.00%       0.000us         0.00%       0.000us       0.000us     100.480us         0.05%     100.480us       2.576us            39  
                                  urEnqueueKernelLaunch        25.92%     131.016ms        25.92%     131.016ms      68.345us       0.000us         0.00%       0.000us       0.000us          1917  
                                       aten::batch_norm         0.18%     901.398us        30.86%     155.956ms       1.218ms       0.000us         0.00%      51.789ms     404.605us           128  
                           aten::_batch_norm_impl_index         0.39%       1.950ms        30.68%     155.055ms       1.211ms       0.000us         0.00%      51.789ms     404.605us           128  
                                            aten::empty         1.65%       8.353ms         1.65%       8.353ms       5.932us       0.000us         0.00%       0.000us       0.000us          1408  
                                       aten::empty_like         0.30%       1.533ms         0.85%       4.300ms      16.798us       0.000us         0.00%       0.000us       0.000us           256  
                                    aten::empty_strided         0.55%       2.767ms         0.55%       2.767ms      10.809us       0.000us         0.00%       0.000us       0.000us           256  
                                            aten::zeros         0.99%       4.995ms        31.88%     161.126ms     209.799us       0.000us         0.00%      69.647ms      90.686us           768  
                                            aten::zero_         0.66%       3.341ms        29.92%     151.189ms     196.861us       0.000us         0.00%      69.647ms      90.686us           768  
autograd::engine::evaluate_function: NativeBatchNorm...         0.71%       3.569ms        39.82%     201.241ms       1.572ms       0.000us         0.00%      81.636ms     637.779us           128  
                               NativeBatchNormBackward0         0.36%       1.844ms        39.11%     197.672ms       1.544ms       0.000us         0.00%      81.636ms     637.779us           128  
autograd::engine::evaluate_function: torch::autograd...         0.77%       3.914ms        15.69%      79.277ms     206.451us       0.000us         0.00%      65.692ms     171.073us           384  
                        torch::autograd::AccumulateGrad         0.85%       4.300ms        14.91%      75.363ms     196.258us       0.000us         0.00%      65.692ms     171.073us           384  
                                           aten::detach         0.00%      20.926us         0.01%      40.113us      13.371us       0.000us         0.00%       0.000us       0.000us             3  
                                                 detach         0.00%      19.187us         0.00%      19.187us       6.396us       0.000us         0.00%       0.000us       0.000us             3  
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  ------------  
Self CPU time total: 505.392ms
Self XPU time total: 199.218ms
```

**Comment by xytintel** [comment]
Test script for normal case:

```python
import torch
import torch.nn as nn

N, C, H, W = 256, 256, 56, 56
# x = torch.randn(N, C, H, W).xpu().to(memory_format=torch.channels_last).bfloat16()
x = torch.randn(N, C, H, W).xpu().bfloat16()
x.requires_grad = True
bn = nn.BatchNorm2d(C).xpu()

prof_xpu = torch.profiler.profile(
    activities=[
    torch.profiler.ProfilerActivity.CPU,
    torch.profiler.ProfilerActivity.XPU],
)
with prof_xpu:
    for i in range(128):
        output = bn(x)
        output.backward(x)
print(prof_xpu.key_averages(group_by_input_shape=True).table(sort_by="self_xpu_time_total", row_limit=100000))
print(output.dtype)
```
Original backward duration: 100.553ms
Optimized backward: 76.273ms

**Review by yucai-intel — APPROVED** [review]
